### DISCUSSION TITLE
Reading only headers and NOT data

### DISCUSSION THREAD

**Discussion by JonCook** [nom-tam-fits/nom-tam-fits#276] [Q&A]
Hi,

Again more of a question. Is there a way when reading an HDU to NOT read the data but just the HeaderCards?

Looking at `Fits.java` - all paths lead to `readHDU` which ultimately reads the data:
```
data.read(this.dataStr);
```
Perhaps it sounds strange, but we are not always interested in reading the data. Sometimes just the header cards and we have some very large products with multidimensional arrays inside.

It would be nice to be able to skip this part if desired. If it's not possible. I'm wondering if there is an easy workaround?

I'm asking as I have a very simple example:
```
Fits fitsSrc = new Fits(filename)
nHDUs = fitsSrc.read().length;
```

In the Fits I have, this read method seems to read indefinitely but not on all platforms. I'm not sure why.

I understand the read method will read the Fits file to the end. I.e, not in deferred mode. Is there an easier/more efficient way to get the number of extensions upfront? Keep calling `getHDU` until it returns null?

Thanks
Jonathan

**Comment by attipaci** [comment]
Hi Jonathan,

I think your question may be partially answered by [README.md](https://github.com/nom-tam-fits/nom-tam-fits#deferred-reading):

> When FITS data are being read from a non-compressed file (`FitsFile`), the `read()` call will parse all HDU headers but will typically skip over the data segments (noting their position in the file however). Only when the user tries to access data from a HDU, will the library load that data from the previously noted file position. The behavior allows to inspect the contents of a FITS file very quickly even when the file is large, and reduces the need for IO when only parts of the whole are of interest to the user. Deferred input, however, is not possible when the input is compressed or if it is uses an stream rather than a random-access `FitsFile`.

I do agree that method naming of `read()` is misleading here, something like `scan()` would be more appropriate, but that is what we have for historical reasons. Nevertheless `read()` should be doing exactly what you describe, when possible. In fact, `read()` simply calls `Fits.readHDU()` until `null` is returned -- just as you wanted to do -- skipping the data themselves if possible.

As for when you can or cannot skip data, it depends on what's the object you use for reading. If it is a `FitsFile`, or some other `RandomAccess` implementation, then `read()` will skip data because you know that you can go back and grab it later. However, if it is a stream that does not support rewinding, you are going to be stuck with reading all the data, whether you like it or not, since you no longer have the option of going back to it at a later point.

That said, I'm not sure why it ends up reading indefinitely on some platforms. It may be something that is worth investigating, as it may potentially be a bug of sorts. If you can find out what triggers and indefinite read and how exactly, we can figure out what can be done to make it better.

-- A.

**Comment by JonCook** [comment]
Hi @attipaci 

Thanks for the reply :-). Sorry, I was a bit hasty when opening this issue. I've changed our parsing to use deferred reading as per the documentation and lazy load the method which counts the total number of HDUs. I've also moved away from stream-based reading as our FITS are not compressed and the problem has gone away.

However, I'm going to try to create a test case that demonstrates this problem to see if there is anything weird going on there with streams and quite large FITS files (300mb+)

Thanks again
Jonathan
